cooperator/tests/test_cooperator.py

```diff
@@ -549,3 +549,79 @@ def test_subscription_request_journal_per_company(self):
         self.assertEqual(
             subscription_request_2.get_journal(), company_2.subscription_journal_id
         )
+
+    def test_create_subscription_request_with_share_of_different_company(self):
+        """
+        Test that creating a subscription request with a share (product)
+        linked to a different company than the current one fails.
+        """
+        # link the share to a specific company (instead of it being shared
+        # across all of them).
+        self.share_y.company_id = self.company.id
+        company_2 = self.create_company("company 2")
+        with self.assertRaises(UserError):
+            self.env["subscription.request"].with_company(company_2).create(
+                self.get_dummy_subscription_requests_vals()
+            )
+
+    def test_create_operation_request_with_share_of_different_company(self):
+        """
+        Test that creating an operation request with a share (product)
+        linked to a different company than the current one fails.
+        """
+        # link the share to a specific company (instead of it being shared
+        # across all of them).
+        self.share_y.company_id = self.company.id
+        company_2 = self.create_company("company 2")
+        vals = {
+            "partner_id": self.demo_partner.id,
+            "operation_type": "sell_back",
+            "share_product_id": self.share_y.id,
+            "quantity": 2,
+        }
+        with self.assertRaises(UserError):
+            self.env["operation.request"].with_company(company_2).create(vals)
+        vals.update(
+            {
+                "share_product_id": self.share_x.id,
+                "share_to_product_id": self.share_y.id,
+            }
+        )
+        with self.assertRaises(UserError):
+            self.env["operation.request"].with_company(company_2).create(vals)
+
+    def test_create_cooperator_for_other_company(self):
+        """
+        Test that creating a cooperator for a different company works and
+        keeps data correctly separated per company.
+        """
+        company_2 = self.create_company("company 2")
+        subscription_request = (
+            self.env["subscription.request"]
+            .with_company(company_2)
+            .create(self.get_dummy_subscription_requests_vals())
+        )
+        self.validate_subscription_request_and_pay(subscription_request)
+        invoice = subscription_request.capital_release_request
+        self.assertEqual(invoice.company_id, company_2)
+
+        partner = subscription_request.partner_id
+        self.assertFalse(partner.coop_candidate)
+        self.assertTrue(partner.member)
+        self.assertTrue(partner.share_ids)
```

Comment:
it is indeed incomplete, since this branch does not yet handle the per-company cooperator data. currently, the partner properties are the same for all companies, so they are correct for some but wrong for others.